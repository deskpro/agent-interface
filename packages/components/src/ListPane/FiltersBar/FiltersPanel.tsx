import * as React from "react";
import classNames from "classnames";

export default React.forwardRef<HTMLDivElement, React.ComponentProps<"div">>(
  ({ className, ...props }, ref) => (
    <div
      ref={ref}
      className={classNames("dp-Filters-panel", className)}
      {...props}
    />
  )
);
