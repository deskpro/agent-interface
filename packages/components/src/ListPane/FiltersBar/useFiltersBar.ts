import { useState, useCallback } from "react";

export type ViewModeType = "list" | "table" | "kanban";

export default () => {
  const [opened, setOpened] = useState<boolean>(false);
  const [locked, setLocked] = useState<boolean>(false);
  const [viewMode, onModeChange] = useState<ViewModeType>("list");

  const onToggleOpenState = useCallback(() => {
    if (opened && locked) {
      setLocked(false);
    }

    setOpened(!opened);
  }, [locked, opened]);

  const onToggleLock = useCallback(() => {
    setLocked(!locked);
  }, [locked, setLocked]);

  return {
    opened,
    onToggleOpenState,
    locked,
    onToggleLock,
    viewMode,
    onModeChange
  };
};
