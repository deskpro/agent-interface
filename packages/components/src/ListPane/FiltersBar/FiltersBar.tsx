import * as React from "react";
import classNames from "classnames";
import { Manager, Reference, Popper } from "react-popper";

import Icon from "../../elements/Icon/Icon";
import { ViewModeType } from "./useFiltersBar";
import FiltersPanel from "./FiltersPanel";

import "@deskpro/agent-interface-style/dist/components/dp-FilterList.css";
import useOutsideClick from "../../hooks/useOutsideClick";

export type FiltersBarSubComponents = {
  Panel: typeof FiltersPanel;
};

export type FiltersBarProps = {
  opened?: boolean;
  locked?: boolean;
  viewMode?: ViewModeType;
  onToggleOpenState: () => void;
  onToggleLock: () => void;
  onModeChange: (nextMode: ViewModeType) => void;
  title: string;
  itemsCount: number;
};

const FiltersBar: React.FC<FiltersBarProps> & FiltersBarSubComponents = ({
  opened = false,
  locked = false,
  viewMode = "list",
  onToggleOpenState,
  onToggleLock,
  onModeChange,
  title,
  itemsCount,
  children
}) => {
  const [barRect, setBarRect] = React.useState<ClientRect>({
    left: 0,
    bottom: 0,
    top: 0,
    right: 0,
    height: 0,
    width: 0
  });
  const barRef = React.useRef<HTMLDivElement>(null);
  React.useEffect(() => {
    if (barRef.current) {
      setBarRect(barRef.current.getBoundingClientRect());
    }
  }, [barRef, setBarRect]);

  const hidePanelOnOutsideClick = React.useCallback(() => {
    if (!locked && opened) {
      onToggleOpenState();
    }
  }, [locked, opened, onToggleOpenState]);
  useOutsideClick(barRef, hidePanelOnOutsideClick);

  return (
    <Manager>
      <div className="dp-Filters" ref={barRef}>
        <div className="dp-Filters-bar dp-Level Level-justify-between">
          <Reference>
            {({ ref }) => (
              <div
                ref={ref}
                className={classNames("dp-Filters-header", {
                  "is-opened": opened,
                  "is-locked": locked
                })}
              >
                <Icon name="menu" size={25} />
                <span className="dp-Filters-title">
                  {title} <span className="dp-qt">({itemsCount})</span>
                </span>
                <Icon
                  name={opened ? "caret-up" : "caret-down"}
                  size={18}
                  color="primary"
                  onClick={onToggleOpenState}
                />
                {opened && (
                  <Icon
                    name="lock"
                    size={18}
                    color="primary"
                    onClick={onToggleLock}
                  />
                )}
              </div>
            )}
          </Reference>

          <ul className="dp-ViewModes">
            <li
              className={classNames("dp-ViewModes-item", {
                "is-active": viewMode === "list"
              })}
            >
              <Icon
                name="list-view"
                size={18}
                onClick={() => onModeChange("list")}
              />
            </li>
            <li
              className={classNames("dp-ViewModes-item", {
                "is-active": viewMode === "table"
              })}
            >
              <Icon
                name="table-view"
                size={18}
                onClick={() => onModeChange("table")}
              />
            </li>
            <li
              className={classNames("dp-ViewModes-item", {
                "is-active": viewMode === "kanban"
              })}
            >
              <Icon
                name="kanban-view"
                size={18}
                onClick={() => onModeChange("kanban")}
              />
            </li>
          </ul>
        </div>

        {opened && !locked && (
          <Popper
            placement="bottom-start"
            modifiers={{
              preventOverflow: {
                enabled: true,
                escapeWithReference: true,
                boundariesElement: "viewport"
              }
            }}
          >
            {({ ref, style }) => (
              <FiltersPanel
                ref={ref}
                style={{
                  height: `calc(100vh - ${barRect.top + barRect.height + 5}px)`,
                  ...style
                }}
              >
                {children}
              </FiltersPanel>
            )}
          </Popper>
        )}
      </div>
    </Manager>
  );
};

FiltersBar.Panel = FiltersPanel;

export default FiltersBar;
