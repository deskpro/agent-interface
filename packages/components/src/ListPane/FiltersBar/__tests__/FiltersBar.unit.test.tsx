import * as React from "react";
import { render, fireEvent } from "react-testing-library";

import FiltersBar from "../FiltersBar";

describe("<FiltersBar />", () => {
  it("should match snapshot for default state", () => {
    const { container } = render(
      <FiltersBar
        title="All Tickets"
        itemsCount={90}
        onToggleOpenState={jest.fn()}
        onToggleLock={jest.fn()}
        onModeChange={jest.fn()}
      />
    );

    expect(container.firstChild).toMatchSnapshot("Initial State");
  });

  it("should match snapshot for opened panel", () => {
    const { container } = render(
      <FiltersBar
        opened
        title="All Tickets"
        itemsCount={90}
        onToggleOpenState={jest.fn()}
        onToggleLock={jest.fn()}
        onModeChange={jest.fn()}
      />
    );

    expect(container.firstChild).toMatchSnapshot("With opened panel");
    expect(container.querySelector(".dp-Filters-panel")).toBeInTheDocument();
  });

  it("should toggle view modes", () => {
    const handleViewModeChange = jest.fn();
    const { container } = render(
      <FiltersBar
        opened
        title="All Tickets"
        itemsCount={90}
        onToggleOpenState={jest.fn()}
        onToggleLock={jest.fn()}
        onModeChange={handleViewModeChange}
      />
    );

    fireEvent.click(container.querySelector(".Icon--table-view"));
    expect(handleViewModeChange).toBeCalledWith("table");
    fireEvent.click(container.querySelector(".Icon--kanban-view"));
    expect(handleViewModeChange).toBeCalledWith("kanban");
  });
});
