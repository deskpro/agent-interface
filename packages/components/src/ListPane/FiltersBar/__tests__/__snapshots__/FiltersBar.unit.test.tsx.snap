// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FiltersBar /> should match snapshot for default state: Initial State 1`] = `
<div
  class="dp-Filters"
>
  <div
    class="dp-Filters-bar dp-Level Level-justify-between"
  >
    <div
      class="dp-Filters-header"
    >
      <span
        class="dp-Icon Icon--menu"
        role="button"
        tabindex="-1"
      >
        <icon-mock
          height="25"
          width="25"
        />
      </span>
      <span
        class="dp-Filters-title"
      >
        All Tickets
         
        <span
          class="dp-qt"
        >
          (
          90
          )
        </span>
      </span>
      <span
        class="dp-Icon Icon--caret-down Icon--primary"
        role="button"
        tabindex="-1"
      >
        <icon-mock
          height="18"
          width="18"
        />
      </span>
    </div>
    <ul
      class="dp-ViewModes"
    >
      <li
        class="dp-ViewModes-item is-active"
      >
        <span
          class="dp-Icon Icon--list-view"
          role="button"
          tabindex="-1"
        >
          <icon-mock
            height="18"
            width="18"
          />
        </span>
      </li>
      <li
        class="dp-ViewModes-item"
      >
        <span
          class="dp-Icon Icon--table-view"
          role="button"
          tabindex="-1"
        >
          <icon-mock
            height="18"
            width="18"
          />
        </span>
      </li>
      <li
        class="dp-ViewModes-item"
      >
        <span
          class="dp-Icon Icon--kanban-view"
          role="button"
          tabindex="-1"
        >
          <icon-mock
            height="18"
            width="18"
          />
        </span>
      </li>
    </ul>
  </div>
</div>
`;

exports[`<FiltersBar /> should match snapshot for opened panel: With opened panel 1`] = `
<div
  class="dp-Filters"
>
  <div
    class="dp-Filters-bar dp-Level Level-justify-between"
  >
    <div
      class="dp-Filters-header is-opened"
    >
      <span
        class="dp-Icon Icon--menu"
        role="button"
        tabindex="-1"
      >
        <icon-mock
          height="25"
          width="25"
        />
      </span>
      <span
        class="dp-Filters-title"
      >
        All Tickets
         
        <span
          class="dp-qt"
        >
          (
          90
          )
        </span>
      </span>
      <span
        class="dp-Icon Icon--caret-up Icon--primary"
        role="button"
        tabindex="-1"
      >
        <icon-mock
          height="18"
          width="18"
        />
      </span>
      <span
        class="dp-Icon Icon--lock Icon--primary"
        role="button"
        tabindex="-1"
      >
        <icon-mock
          height="18"
          width="18"
        />
      </span>
    </div>
    <ul
      class="dp-ViewModes"
    >
      <li
        class="dp-ViewModes-item is-active"
      >
        <span
          class="dp-Icon Icon--list-view"
          role="button"
          tabindex="-1"
        >
          <icon-mock
            height="18"
            width="18"
          />
        </span>
      </li>
      <li
        class="dp-ViewModes-item"
      >
        <span
          class="dp-Icon Icon--table-view"
          role="button"
          tabindex="-1"
        >
          <icon-mock
            height="18"
            width="18"
          />
        </span>
      </li>
      <li
        class="dp-ViewModes-item"
      >
        <span
          class="dp-Icon Icon--kanban-view"
          role="button"
          tabindex="-1"
        >
          <icon-mock
            height="18"
            width="18"
          />
        </span>
      </li>
    </ul>
  </div>
  <div
    class="dp-Filters-panel"
    style="position: absolute; top: 0px; left: 0px; opacity: 0; pointer-events: none;"
  />
</div>
`;
